pyfixest/estimation/feiv_.py

```diff
@@ -193,3 +206,225 @@ def get_fit(self) -> None:
         # Compute bread matrix
         D = np.linalg.inv(self._tXZ @ self._tZZinv @ self._tZX)
         self._bread = H.T @ D @ H
+
+    def first_stage(self) -> None:
+        """Implement First stage regression."""
+        from pyfixest.estimation.estimation import feols
+
+        #  Start First Stage
+        _X = self._X
+        _Z = self._Z
+        _Y = self._Y
+
+        # Store names of instruments from Z matrix
+        self._non_exo_instruments = list(set(self._coefnames_z) - set(self._coefnames))
+
+        # Prepare your data
+        # Select the columns from self._data that match the variable names
+        data = pd.DataFrame(self._Z_1st_stage, columns=self._coefnames_z)
+
+        # Store instrument variable matrix for futyre use
+        _Z_iv = data[self._non_exo_instruments]
+        self._Z_iv = _Z_iv.to_numpy()
+        # Dynamically create the formula string
+        # Extract names of fixed effect
+
+        data["Y"] = self._endogvar_1st_stage
+        independent_vars = " + ".join(
+            data.columns[:-1]
+        )  # All columns except the last one ('Y')
+
+        # Set fix effects, cluster, and weight options to be passed to feols.
+
+        if self._has_fixef:
+            FE_vars = self._fixef.split("+")
+            data[FE_vars] = self._data[FE_vars]
+            formula = f"Y ~ {independent_vars} | {self._fixef}"
+        else:
+            formula = f"Y ~ {independent_vars}"
+
+        # Type hint to reflect that vcov_detail can be either a dict or a str
+        vcov_detail: Union[dict[str, str], str]
+
+        if self._is_clustered:
+            a = self._clustervar[0]
+            vcov_detail = {self._vcov_type_detail: a}
+            data[a] = self._data[a]
+        else:
+            vcov_detail = self._vcov_type_detail
+
+        if self._has_weights:
+            data["weights"] = self._weights
+            weight_detail = "weights"
+        else:
+            weight_detail = None
+
+        # Do first stage regression
+
+        model1 = feols(
+            fml=formula,
+            data=data,
+            vcov=vcov_detail,
+            weights=weight_detail,
+            collin_tol=1e-10,
+        )
+        # Ensure model1 is of type Feols
+        if isinstance(model1, Feols):
+            # Store the first stage coefficients
+            self._pi_hat = model1._beta_hat
+
+            # Use fitted values from the first stage
+            self._X_hat = model1._Y_hat_link
+
+            # Residuals from the first stage
+            self._v_hat = model1._u_hat
+
+            # Store 1st stage model for further use
+            self._model_1st_stage = model1
+        else:
+            raise TypeError("The first stage model must be of type Feols")
+
+    def IV_Diag(self) -> None:
+        """Implement IV weakness test(F-test)."""
+        # This method covers hetero-robust and clustered-robust F stats
```

Comment:
Could you move the description of the method into the string above? We're using the `numpy` format, e.g. the docstring could look like this: 

```python
def IV_weakness_test():
    """
    Implement IV weakness test (F-test).

    This method covers hetero-robust and clustered-robust F statistics. It produces two statistics:
    
    - self._f_stat_1st_stage: F statistics of first stage regression
    - self._eff_F: Effective F statistics (Olea and Pflueger 2013) of first stage regression
    
    Notes
    -----
    "self._f_stat_1st_stage" is adjusted to the specification of vcov. If vcov_detail = "iid", F statistics is not adjusted,
    otherwise it is always adjusted.
    
    Example
    -------
    The following is an example usage of this method:
    
        ```{python}
   
        import numpy as np
        import pandas as pd
        from pyfixest.estimation.estimation import feols

        # Set random seed for reproducibility
        np.random.seed(1)

        # Number of observations
        n = 1000

        # Simulate the data
        # Instrumental variable
        z = np.random.binomial(1, 0.5, size=n)
        z2 = np.random.binomial(1, 0.5, size=n)

        # Endogenous variable
        d = 0.5 * z + 1.5 * z2 + np.random.normal(size=n)

        # Control variables
        c1 = np.random.normal(size=n)
        c2 = np.random.normal(size=n)

        # Outcome variable
        y = 1.0 + 1.5 * d + 0.8 * c1 + 0.5 * c2 + np.random.normal(size=n)

        # Cluster variable
        cluster = np.random.randint(1, 50, size=n)
        weights = np.random.uniform(1, 3, size=n)

        # Create a DataFrame
        data = pd.DataFrame({
            'd': d,
            'y': y,
            'z': z,
            'z2': z2,
            'c1': c1,
            'c2': c2,
            'cluster': cluster,
            'weights': weights
        })

        vcov_detail = "iid"
        
        # Fit OLS model
        fit_ols = feols("y ~ 1 + d + c1 + c2", data=data, vcov=vcov_detail)

        # Fit IV model
        fit_iv = feols("y ~ 1 + c1 + c2 | d ~ z", data=data, vcov=vcov_detail, weights="weights")
        fit_iv.IV_Diag()

        print("(Unadjusted) F stat :", fit_iv._f_stat_1st_stage)
        print("Effective F stat :", fit_iv._eff_F)

        # The example above generates the following results
        # (Unadjusted) F stat : 52.81535560457482
        # Effective F stat : 48.661542741328205
   ```
    """
    pass
   
```

Reply:
This layout looks far great! Yes, we also have to follow the numpy docstring. I will place the new method `IV_weakness_test()` as a separate method, and put the documentation as you suggested!